### PR TITLE
Creating Rspec tests for app

### PR DESCRIPTION
**Pull request by jmarsh24**
This branch is dedicated to creating all the app tests

### CODE REVIEW

File: spec/models/video_spec.rb [hunk 1]
```diff
@@ -46,5 +46,724 @@
...
+    video.clicked!
+    expect(video.click_count).to eq(1)
+  end
+
+  describe ".display.any_song_attribute" do
+    it "only song exists" do
+      song = create(:song, title: "No Vendrá", artist: "Angel D'AGOSTINO", genre: "TANGO")
+      video = create(:video, click_count: 0, song: song)
+
+      expect(video.display.any_song_attributes).to eq("No Vendrá - Angel D'Agostino - Tango")
```
Comment (thomasstephane) [inline spec/models/video_spec.rb:71]: Don't test the video test the display, so your tests in the display space should be like that:

```ruby
# in spec/models/video/display_spec.rb
Rspec.describe Video::Display do
  describe "#any_song_attributes" do
    context "when only the song exists" do
      it "returns the song name" do
        song = build(:song, title: "No Vendrá", artist: "Angel D'AGOSTINO", genre: "TANGO")
        video = build(:video, song: song)
        display = described_class.new(video)
        expect(display.any_song_attributes).to eq "No Vendrá - Angel D'Agostino - Tango"
      end
    end
  end
end
```

File: spec/models/video_spec.rb [hunk 2]
```diff
@@ -46,5 +46,724 @@
...
+      expect(video.display.dancer_names).to eq(nil)
+    end
+
+    it "has both leader and follower" do
+      leader = create(:leader, name: "Carlitos Espinoza")
+      follower = create(:follower, name: "Noelia Hurtado")
+      video = create(:video, leader: leader, follower: follower)
+      expect(video.display.dancer_names).to eq("Carlitos Espinoza & Noelia Hurtado")
+    end
+  end
```
Comment (thomasstephane) [inline spec/models/video_spec.rb:207]: Same for all these display things

File: spec/models/video_spec.rb [hunk 3]
```diff
@@ -46,5 +46,724 @@
...
+
+  describe ".filter_by_not_hidden" do
+    it "filter_by_not_hidden" do
+      video = create(:video, hidden: false)
+      @result = described_class.not_hidden
+      expect(@result).to include(video)
+    end
+  end
```
Comment (thomasstephane) [inline spec/models/video_spec.rb:300]: All these scopes should be under a `describe "scope" do` block to be easier to keep together, also same comment as before, you aren't testing anything much here :)

File: spec/models/video_spec.rb [hunk 4]
```diff
@@ -46,5 +46,724 @@
...
+
+  describe ".paginate" do
```
Comment (thomasstephane) [inline spec/models/video_spec.rb:302]: remove this if you don't test it

File: spec/models/video_spec.rb [hunk 5]
```diff
@@ -46,5 +46,724 @@
...
+  end
+
+  describe ".has_song" do
+    it "returns videos with a song" do
+      song = create(:song)
+      video = create(:video, song: song)
+      @result = described_class.has_song
+      expect(@result).to include(video)
```
Comment (thomasstephane) [inline spec/models/video_spec.rb:310]: same here, useless test, ensure that things that shouldn't be included aren't

File: spec/models/video_spec.rb [hunk 6]
```diff
@@ -46,5 +46,724 @@
...
+    end
+
+    it "returns video with song artist that matches query" do
+      song = create(:song, artist: "Angel D'Agostino")
+      video = create(:video, song: song)
+      VideosSearch.refresh
+      @result = described_class.filter_by_query("d'agostino")
+      expect(@result).to include(video)
+    end
+  end
```
Comment (thomasstephane) [inline spec/models/video_spec.rb:768]: And same for everything else here

File: spec/models/videos_search_spec.rb [hunk 7]
```diff
@@ -8,5 +8,19 @@
 require "rails_helper"
 
 RSpec.describe VideosSearch, type: :model do
-  pending "add some examples to (or delete) #{__FILE__}"
+  describe ".refresh" do
+    it "materialized view able to be refreshed" do
+      video = create(:video)
+      VideosSearch.refresh
+      expect(VideosSearch.count).to eq(1)
```
Comment (thomasstephane) [inline spec/models/videos_search_spec.rb:15]: what is this testing? Maybe you want to test that the refresh change the count? if so:

```ruby
describe ".refresh" do
  it "includes new video when refreshed" do
    create(:video)
    expect { described_class.refresh }.to change(described_class, :count).from(0).to(1)
  end
end
```

File: spec/models/videos_search_spec.rb [hunk 8]
```diff
@@ -8,5 +8,19 @@
...
+    end
+  end
+
+  describe ".search" do
+    it "materialized view is searchable" do
+      video = create(:video)
+      VideosSearch.refresh
+      expect(VideosSearch.count).to include(1)
```
Comment (thomasstephane) [inline spec/models/videos_search_spec.rb:23]: this is not testing search, this is a copy paste from above

File: spec/models/channel_spec.rb [hunk 9]
```diff
@@ -17,6 +17,28 @@
 require "rails_helper"
 
 RSpec.describe Channel, type: :model do
+  let(:channel) { build(:channel) }
```
Comment (thomasstephane) [inline spec/models/channel_spec.rb:20]: no.

File: spec/models/channel_spec.rb [hunk 10]
```diff
@@ -17,6 +17,28 @@
...
+    it "changes imported to true if total_videos_count is smaller than internal videos_count" do
+      channel = create(:channel, imported: false, total_videos_count: 500, videos_count: 501)
+      channel.save
+      expect(channel.reload.imported).to eq(true)
+    end
+
+    it "changes imported to false if total_videos_count is greater than internal videos_count" do
+      channel = create(:channel, imported: false, total_videos_count: 500, videos_count: 500)
+      channel.save
+      expect(channel.reload.imported).to eq(true)
```
Comment (thomasstephane) [inline spec/models/channel_spec.rb:41]: none of these tests is testing the `update_videos_count` unless I'm missing something

File: spec/models/song_spec.rb [hunk 11]
```diff
@@ -1,7 +1,132 @@
...
+      song = create(:song, title: "Tal vez será su voz", artist: "Anibal Troilo", genre: "TANGO")
+      expect(song.full_title).to eq("Tal Vez Será Su Voz - Anibal Troilo - Tango")
+    end
+  end
+
+  describe ".set_all_popularity_values" do
+    it "calculates popularity values" do
+      song = create(:song, popularity: 0)
+      described_class.set_all_popularity_values
+      expect(song.reload.popularity).to eq(100)
```
Comment (thomasstephane) [inline spec/models/song_spec.rb:129]: this normalizes right? you should have more than one song then

File: spec/factories/channels.rb [hunk 12]
```diff
@@ -17,15 +17,6 @@
...
-    title { "MyText" }
-    imported { true }
-    thumbnail_url { "MyText" }
-  end
-
-  factory :random_channel do
-    title { Faker::Name.first_name }
-    thumbnail_url { Faker::Name.last_name }
-    name { "#{Faker::Name.first_name} #{Faker::Name.last_name}" }
+    channel_id { Faker::String.random(length: 12) }
```
Comment (thomasstephane) [inline spec/factories/channels.rb:20]: Use a sequence here. https://github.com/thoughtbot/factory_bot/blob/master/GETTING_STARTED.md#sequences

File: app/models/song.rb [hunk 13]
```diff
@@ -45,13 +44,16 @@ class Song < ApplicationRecord
...
   def full_title
-    "#{title.titleize} - #{artist.titleize} - #{genre.titleize}"
+    "#{title.titleize} - #{artist.split("'").map(&:titleize).join("'")} - #{genre.titleize}"
   end
 
   class << self
     def set_all_popularity_values
-      max_count = Song.maximum(:videos_count).to_f
-      update_all("popularity = CAST(videos_count AS FLOAT) / #{max_count} * 100.0")
+      update_all("popularity = CAST(videos_count AS FLOAT) / #{max_times_song_appears} * 100.0")
```
Comment (thomasstephane) [inline app/models/song.rb:52]: NEVER use update_all. It bypasses validations. This is a horrible method you should never ever use. Iterate over the values instead, yes it's slower, that's ok.

File: app/models/song.rb [hunk 14]
```diff
@@ -45,13 +44,16 @@ class Song < ApplicationRecord
...
+    end
+
+    def max_times_song_appears
+      @max_times_song_appears ||= Song.maximum(:videos_count).to_f
```
Comment (thomasstephane) [inline app/models/song.rb:56]: the `to_f` is unnecessary

File: app/models/channel.rb [hunk 15]
```diff
@@ -15,29 +15,31 @@
...
-    def update_import_status
-      where("imported_videos_count < total_videos_count").find_each do |channel|
-        channel.update(imported: false)
-      end
+  def update_imported
+    self.imported = videos_count >= total_videos_count
+  end
+
+  class << self
+    def imported?
```
Comment (thomasstephane) [inline app/models/channel.rb:41]: the method name is misleading, method with `?` should be just testing if something is true or false, not updating it.

File: app/models/channel.rb [hunk 16]
```diff
@@ -15,29 +15,31 @@
...
+      where("videos_count < total_videos_count").update_all(imported: false)
```
Comment (thomasstephane) [inline app/models/channel.rb:42]: update_all bypass validations. DO not use it.

File: app/models/channel.rb [hunk 17]
```diff
@@ -15,29 +15,31 @@
...
+  before_save :update_imported, if: -> { total_videos_count_changed? || videos_count_changed? }
+
   include Reviewable
   include Importable
 
   validates :channel_id, presence: true, uniqueness: true
 
   has_many :videos, dependent: :destroy
 
+  scope :imported, -> { where("`channels`.`videos_count` >= `channels`.`total_videos_count`") }
```
Comment (thomasstephane) [inline app/models/channel.rb:27]: you don't need the ` everywhere here.

```where("channels.videos_count >= channels.total_videos_count")``` should work

File: app/models/channel.rb [hunk 18]
```diff
@@ -15,29 +15,31 @@
 #  reviewed              :boolean          default(FALSE)
 #
 class Channel < ApplicationRecord
+  before_save :update_imported, if: -> { total_videos_count_changed? || videos_count_changed? }
```
Comment (thomasstephane) [inline app/models/channel.rb:18]: what about a method for your if that check both of these? Avoid inline blocks for before saves it makes it hard to read 

```ruby
before_save :update_imported, if: :count_changed?
```

**Review by thomasstephane — CHANGES_REQUESTED** [review]

File: spec/models/song_spec.rb [hunk 19]
```diff
@@ -1,7 +1,124 @@
...
+      song = create(:song, active: false)
+      expect(described_class.filter_by_not_active).to include(song)
+    end
+  end
+
+  describe ".title_match" do
+    it "find song by title" do
+      song = create(:song, title: "No Vendrá")
+      @result = described_class.title_match("No Vendrá")
+      expect(@result).to include(song)
```
[the remark this answers is not in the document]
Comment (jmarsh24) [reply spec/models/song_spec.rb]: Do you mean I should create a separate song with a title that doesn't match and make sure that it doesn't return that song?

`    it "find song by title" do
      song = create(:song, title: "No Vendrá")
      song2 = create(:song, title: "Not matching song")
      result = described_class.title_match("no vendra")
      expect(result).to include(song)
      expect(result).not_to include(song2)
    end`
    
    Something like this?

File: spec/models/videos_search_spec.rb (hunk 7 above)
Comment (jmarsh24) [reply spec/models/videos_search_spec.rb:15]: Do you have any idea why this is happening?

image

File: spec/models/channel_spec.rb (hunk 10 above)
Comment (jmarsh24) [reply spec/models/channel_spec.rb:49]: Channel has a before save action so upon reload the channel imported_value should be change to the correct imported state based upon the total_video_count and videos_count

Comment (jmarsh24) [reply spec/models/channel_spec.rb:49]: I'm not sure if it's the correct way to test it though. 🤔 but it's simple and seems to work.

File: spec/models/video_spec.rb (hunk 3 above)
Comment (jmarsh24) [reply spec/models/video_spec.rb:203]: Is it ok to use describe twice? or should I use context scope, describe scope....?

File: spec/models/video_spec.rb [hunk 20]
```diff
@@ -43,8 +43,760 @@
...
+        create_list(:video, 200)
+        page1 = described_class.paginate(1, 120)
+        page2 = described_class.paginate(2, 120)
+        page3 = described_class.paginate(3, 120)
+        expect(page1.count).to eq(120)
+        expect(page2.count).to eq(80)
+        expect(page3.count).to eq(0)
+      end
+    end
+  end
```
Comment (thomasstephane) [inline spec/models/video_spec.rb:801]: run rubicon on this file (and all the others....) plenty to correct here :)

File: spec/support/models/shared_examples/importable.rb [hunk 21]
```diff
@@ -0,0 +1,10 @@
+RSpec.shared_examples "an importable" do |factory_name|
+  describe ".imported" do
+    it "includes objects with imported flagged" do
+      imported = create(factory_name, imported: true)
+      not_imported = create(factory_name, imported: false)
+      expect(described_class.imported).to include(imported)
+      expect(described_class.imported).not_to include(not_imported)
```
Comment (thomasstephane) [inline spec/support/models/shared_examples/importable.rb:7]: same as before, check an equality instead

**Review by thomasstephane — CHANGES_REQUESTED** [review]

File: app/models/playlist.rb [hunk 22]
```diff
@@ -16,7 +16,6 @@
 #  updated_at    :datetime         not null
 #
 class Playlist < ApplicationRecord
-  include Reviewable
```
[the remark this answers is not in the document]
Comment (jmarsh24) [reply app/models/playlist.rb:19]: if the title is less than 2 strings long, then there will be a lot of false positives.

File: app/models/event.rb [hunk 23]
```diff
@@ -21,28 +21,39 @@ class Event < ApplicationRecord
...
+                  unaccent(videos.tags) ILIKE unaccent(:query) OR
+                  unaccent(channels.title) ILIKE unaccent(:query)",
+                query: "%#{search_title}%")
+  end
 
-    def match_event(event_id)
-      event = Event.find(event_id)
-      event_title = event.title.split("-")[0].strip
+  def match_videos
+    return if search_title.split.size < 2 || videos_with_event_title_match.empty?
```
[the remark this answers is not in the document]
Comment (jmarsh24) [reply app/models/event.rb:40]: if the title is less than 2 strings long, then there will be a lot of false positives.

File: app/models/playlist.rb (hunk 22 above)
[the remark this answers is not in the document]
Comment (jmarsh24) [reply app/models/playlist.rb:19]: I didn't have a column in the database for it.

File: app/models/song.rb [hunk 24]
```diff
@@ -24,34 +24,31 @@ class Song < ApplicationRecord
...
-                              where("unaccent(songs.title) ILIKE unaccent(:query) OR
-                                    unaccent(artist) ILIKE unaccent(:query) OR
-                                    unaccent(genre) ILIKE unaccent(:query)",
-                                    query: "%#{query}%").references(:song)
-                            }
+  scope :title_match, ->(query) { where("unaccent(title) ILIKE unaccent(?)", "%#{query}%") }
 
   def full_title
-    "#{title.titleize} - #{artist.titleize} - #{genre.titleize}"
+    "#{title.titleize} - #{artist.split("'").map(&:titleize).join("'")} - #{genre.titleize}"
```
[the remark this answers is not in the document]
Comment (jmarsh24) [reply app/models/song.rb:42]: Titlize doesn't title a name like "D'Agostino" properly it would produce "D'agostino" instead.

**Review by thomasstephane — APPROVED** [review]

**Pull request merged**